### PR TITLE
feat: check bot existence before asking and refactor engine selection

### PR DESCRIPTION
**Pull request by nayyara-airlangga**
This change will check for a bot's existence before asking. I've also refactored our tests and logic for bot selection to implement the Strategy and Factory pattern to decouple the implementation of a chat engine to the bot

### CODE REVIEW

**Comment by TakeMeGH** [comment]
I’m not sure if this task includes a fix related to this issue. But I want to ask if this already handles the errors returned by the Slack app according to the errors raised by the code ?. Because it seems that previously it only displayed the error template from the Slack app.

**Comment by nayyara-airlangga** [comment]
> I’m not sure if this task includes a fix related to this issue. But I want to ask if this already handles the errors returned by the Slack app according to the errors raised by the code ?. Because it seems that previously it only displayed the error template from the Slack app.

@TakeMeGH, which error are you talking about?

**Comment by TakeMeGH** [comment]
> > I’m not sure if this task includes a fix related to this issue. But I want to ask if this already handles the errors returned by the Slack app according to the errors raised by the code ?. Because it seems that previously it only displayed the error template from the Slack app.
> 
> @TakeMeGH, which error are you talking about?

Like when sending a message fails, for example, due to a missing channel ID or perhaps an error with the Slack API when attempting to send the message. The error displayed is just 'dispatch failed.' I'm not sure if this is an issue or not, but the error returned through the Slack app doesn’t seem very representative.

File: adapter/slack.py [hunk 1]
```diff
@@ -1,79 +1,112 @@
...
-                ts=loading_message["ts"],
-                text=chatbot_response
+                channel=channel_id, ts=loading_message["ts"], text=chatbot_response
             )
-            
+
             return Response(status_code=200)
 
+        except sqlalchemy.exc.DataError as e:  # pragma: no cover
+            return {"text": "Whoops. Can't found the chatbot you're looking for"}
```
Comment (Insta-x) [inline adapter/slack.py:76]: I think this error message needs to be differentiated with the one above in line 52. The one in line 52 happens if there really is no bot with the id, while this line happens because of some error. Maybe something like "Database Error" or something.

Comment (nayyara-airlangga) [reply adapter/slack.py:76]: That one is to handle non-UUIDs as the database throws an error when that happens (e.g. 1212121). Since the error we received is quite hard to parse, I used this as a catch-all for now

**Comment by nayyara-airlangga** [comment]
> > > I’m not sure if this task includes a fix related to this issue. But I want to ask if this already handles the errors returned by the Slack app according to the errors raised by the code ?. Because it seems that previously it only displayed the error template from the Slack app.
> > 
> > 
> > @TakeMeGH, which error are you talking about?
> 
> Like when sending a message fails, for example, due to a missing channel ID or perhaps an error with the Slack API when attempting to send the message. The error displayed is just 'dispatch failed.' I'm not sure if this is an issue or not, but the error returned through the Slack app doesn’t seem very representative.

@TakeMeGH for this kind of error, it's something we can't really handle for now. The handling I made was to better relay when a user-caused error happens such as missing parameters or non-existent bot ids

**Comment by jasonirvine76** [comment]
This PR looks great! The refactoring with the Strategy and Factory patterns is a solid improvement, as it cleanly decouples the bot selection from the chat engine, which should make future modifications easier. Nice work!

**Comment by TakeMeGH** [comment]
> > > > I’m not sure if this task includes a fix related to this issue. But I want to ask if this already handles the errors returned by the Slack app according to the errors raised by the code ?. Because it seems that previously it only displayed the error template from the Slack app.
> > > 
> > > 
> > > @TakeMeGH, which error are you talking about?
> > 
> > 
> > Like when sending a message fails, for example, due to a missing channel ID or perhaps an error with the Slack API when attempting to send the message. The error displayed is just 'dispatch failed.' I'm not sure if this is an issue or not, but the error returned through the Slack app doesn’t seem very representative.
> 
> @TakeMeGH for this kind of error, it's something we can't really handle for now. The handling I made was to better relay when a user-caused error happens such as missing parameters or non-existent bot ids

Alright, thank you for the clarification.

**Review by TakeMeGH — APPROVED** [review]

File: adapter/slack.py (hunk 1 above)
Comment (Insta-x) [reply adapter/slack.py:76]: understandable, thanks for the explanation
